### ISSUE TITLE
Case for extending fungible token info

### ISSUE THREAD

**Issue by nathandem** [solana-labs/solana-program-library#4143]
### TL;DR

- Metaplex metadata is the de-facto standard to add basic info to fungible SLP-tokens. Yet, it's a poor design.
- I propose to add an official token program attach a PDA to token mints to store basic info (name, symbol, etc)
- If this proposal doesn't gather enough support, at least, Solana docs about token creation should be updated to include info about the Metaplex metadata.

---------------

### Longer version

Since the end of [token-list](https://github.com/solana-labs/token-list), [Metaplex NFT system](https://docs.metaplex.com/programs/token-metadata/overview#a-json-standard) has slowly become the de-facto standard to add some basic info to fungible SLP-tokens. 

I think it's suboptimal.

Why? It's not fitting, at least for 2 reasons:
- Most data stored on-chain only make sense for NFT collections. Fungible tokens don't need fields like `primarySaleHappened` and `editionNonce`. It's heavy and confusing.
- The required `upgradeAuthority` doesn't play well with token mints that want to remove their minting authority/ability.
- Overall the design is clunky. There's no need to have a PDA with a URI pointing to an off-chain JSON itself pointing to a logo.

Here's an example: https://solscan.io/token/mb1eu7TzEc71KxDpsmsKoucSSuuoGLv1drys1oP2jh6#metadata

Instead, I'd like fungible tokens to include key info:
- name
- symbol
- description
- image

As we can't modify the spl-token `Mint` struct ([link](https://docs.rs/spl-token/3.5.0/spl_token/state/struct.Mint.html)) to avoid causing breaking changes (too late!), I propose to store this data as a PDA to the Token program or the Token-2022. Also the web3.js lib could wrap this operation when devs wish to create new tokens.

Benefits:
- Better developer experience. I personally find this design inelegant. But more importantly, though it might not look much the friction coming from this suboptimal design repeated over thousands of devs become, in the aggregate, a big friction.
- It'll make it easier for explorers to show this information.

There are different ways to go about this. But I'd first like to get the opinion of some other developers and contributors about this before going further.

What do you think?

@joncinque @mvines @CriesofCarrots @2501babe @SebastianBor @ngundotra @jacobcreech @yihau 

OPTION B:

If my proposal doesn't seem compelling enough, we still can't leave the status quo, at least documentation wise. The [SPL-token doc](https://spl.solana.com/token) doesn't include any mention of Metaplex metadata. Same for the [cookbook](https://solanacookbook.com/references/token.html). Both are the trusted go-to docs for new devs on solana and it's not even mentioned!

Let's take my case to illustrate why it's a problem. I minted a fungible token a few days ago. After minting its supply, I set the minting authority to `null` as I wanted it to have a fixed supply. I distributed the token, including in a vesting contract. It's only later when looking at how to add details about the token in Phantom that I discovered about this Metaplex standard ([here](https://docs.phantom.app/best-practices/tokens)).

I did some extra research about it and understood how it works after reading [this tuto on Quicknode's blog](https://www.quicknode.com/guides/solana-development/spl-tokens/how-to-create-a-fungible-spl-token-with-the-new-metaplex-token-standard) (thx @amilz). But given that I already set my minting authority to `null`, it's not possible anymore to create a Metaplex metadata's PDA for this Mint. So I'm left with either a) having a unkown token on wallets and explorers or b) telling eveyone that the token is deprecated, make a snapshot of the balances, mint a new token and the Metaplex's PDA, reallocate like before.

It's not such a big deal for me as I'm still in test phase. But beyond me, I think if we keep the current suboptimal de facto standard of metaplex metadata, at least it should be known/explained to devs looking to create tokens so they know about it upfront.

**Comment by joncinque** [comment]
While I agree that there is a lot more than is strictly required in the token-metadata program, I'll need more clarification on your proposal.

The problem that you ran into was a lack of mint authority to create the metaplex metadata, so you're proposing token-metadata that is *not* the metaplex metadata. As you correctly point out, we can't change the mint structure, so having it in a separate PDA is definitely the right approach. In general, the token program should not manage metadata, since it has nothing to do with the operation of the token protocol, so we need some other protocol, meaning some new program with a PDA that holds token metadata.

How does the metadata program authorize the creation of token metadata? Metaplex's solution was to use the token mint authority, and it was also deemed acceptable in its first iteration in SPL at #1544. Most likely this new program would do the same thing, so you'd still be out of luck unfortunately!

We can certainly roll a slimmed down version of the token-metadata program, but how would this solve your issue? In the end, there would still be a requirement to create the account by signing with the mint authority.

**Comment by nathandem** [comment]
Thanks for the answer @joncinque,

While I agree with you that metadata aren't technically needed for the operations of the token mint itself, they are still required for end users. End users don't want to manipulate unknown tokens in their wallets (that's what Ethereum folks got right when they put the name in the ERC20 standard - though there's no image, it was a bit early for it).

The main reason I'm creating this issue is to say that in most cases we can't get away with not setting metadata on user facing fungible tokens hence setting said metadata should be easy for solana devs.

Primarily that'd mean that it's well documented, in the official documentation in particular. And/or that creating and attaching metadata to tokens is part of the tooling (web3.js and CLI) such that devs know about it and can't make mistakes. Given that it's the factual default, when creating a token mint, devs should have to explicitly opt-out from attaching metadata if they don't want them. (In that regard, the problem I ran into is more a consequence of the lack of documentation and tooling around this metadata. Once this is solved future solana devs shouldn't even find themselves in this situation.)

Second, if this idea is supported, adding a dedicated program to the SPL repo would be a step up for fungible tokens in my opinion. As pointed out early, the Metaplex metadata standard is a) overly complex for fungible tokens and b) all data could fit onchain. Such program could be modeled/in the same vein as the associated token account one, as in it's a light PDA that is attached to the mint. In terms of what data this pda should hold, I'd tend to say name, symbol, image uri and a description. Also having this program be an "official" solana spl program would make it easier to integrate it in the Solana tooling and reach the smooth devX that I'm referring early with the tooling (and doc).

**Comment by joncinque** [comment]
So if I have it right, you're suggesting:

* slimmed down metadata program in SPL, that still requires the mint authority to sign creation
* integrate this program in spl-token tools, ie. CLI and JS, so that metadata can be easily made at the same time as the token mint

Is that about right?

**Comment by nathandem** [comment]
That's exactly that 🎯

**Comment by joncinque** [comment]
The spl-token-metadata program was implemented in token-2022 with #4648, as an interface in  #4514, and an example program in #4546. I'll be adding CLI support within the next couple of weeks

**Comment by joncinque** [comment]
Closing since this all done with #5094 and others
